# JSON examples and exercise
****
+ get familiar with packages for dealing with JSON
+ study examples with JSON strings and files 
+ work on exercise to be completed and submitted 
****
+ reference: http://pandas.pydata.org/pandas-docs/stable/io.html#io-json-reader
****

In [1]:
import pandas as pd
import numpy as np

## imports for Python, Pandas

In [2]:
import json
from pandas.io.json import json_normalize

## JSON example, with string

+ demonstrates creation of normalized dataframes (tables) from nested json string
+ source: http://pandas.pydata.org/pandas-docs/stable/io.html#normalization

In [3]:
# define json string
data = [{'state': 'Florida', 
         'shortname': 'FL',
         'info': {'governor': 'Rick Scott'},
         'counties': [{'name': 'Dade', 'population': 12345},
                      {'name': 'Broward', 'population': 40000},
                      {'name': 'Palm Beach', 'population': 60000}]},
        {'state': 'Ohio',
         'shortname': 'OH',
         'info': {'governor': 'John Kasich'},
         'counties': [{'name': 'Summit', 'population': 1234},
                      {'name': 'Cuyahoga', 'population': 1337}]}]

In [4]:
#json_normalize(data, 'counties', [['info','governor'], 'state','shortname'], errors='ignore')
json_normalize(data, 'counties')

name  population
0        Dade       12345
1     Broward       40000
2  Palm Beach       60000
3      Summit        1234
4    Cuyahoga        1337

In [6]:
# further populate tables created from nested element
json_normalize(data, 'counties', ['state', 'shortname', ['info', 'governor']])

name  population    state shortname info.governor
0        Dade       12345  Florida        FL    Rick Scott
1     Broward       40000  Florida        FL    Rick Scott
2  Palm Beach       60000  Florida        FL    Rick Scott
3      Summit        1234     Ohio        OH   John Kasich
4    Cuyahoga        1337     Ohio        OH   John Kasich

****
## JSON example, with file

+ demonstrates reading in a json file as a string and as a table
+ uses small sample file containing data about projects funded by the World Bank 

In [7]:
# load json as string
json.load((open('data/world_bank_projects_less.json')))

[{'_id': {'$oid': '52b213b38594d8a2be17c780'},
  'approvalfy': 1999,
  'board_approval_month': 'November',
  'boardapprovaldate': '2013-11-12T00:00:00Z',
  'borrower': 'FEDERAL DEMOCRATIC REPUBLIC OF ETHIOPIA',
  'closingdate': '2018-07-07T00:00:00Z',
  'country_namecode': 'Federal Democratic Republic of Ethiopia!$!ET',
  'countrycode': 'ET',
  'countryname': 'Federal Democratic Republic of Ethiopia',
  'countryshortname': 'Ethiopia',
  'docty': 'Project Information Document,Indigenous Peoples Plan,Project Information Document',
  'envassesmentcategorycode': 'C',
  'grantamt': 0,
  'ibrdcommamt': 0,
  'id': 'P129828',
  'idacommamt': 130000000,
  'impagency': 'MINISTRY OF EDUCATION',
  'lendinginstr': 'Investment Project Financing',
  'lendinginstrtype': 'IN',
  'lendprojectcost': 550000000,
  'majorsector_percent': [{'Name': 'Education', 'Percent': 46},
   {'Name': 'Education', 'Percent': 26},
   {'Name': 'Public Administration, Law, and Justice', 'Percent': 16},
   {'Name': 'Educatio

In [8]:
# load as Pandas dataframe
sample_json_df = pd.read_json('data/world_bank_projects_less.json')
sample_json_df

_id  approvalfy board_approval_month  \
0  {'$oid': '52b213b38594d8a2be17c780'}        1999             November   
1  {'$oid': '52b213b38594d8a2be17c781'}        2015             November   

      boardapprovaldate                                 borrower  \
0  2013-11-12T00:00:00Z  FEDERAL DEMOCRATIC REPUBLIC OF ETHIOPIA   
1  2013-11-04T00:00:00Z                    GOVERNMENT OF TUNISIA   

            closingdate                              country_namecode  \
0  2018-07-07T00:00:00Z  Federal Democratic Republic of Ethiopia!$!ET   
1                   NaN                      Republic of Tunisia!$!TN   

  countrycode                              countryname countryshortname  ...  \
0          ET  Federal Democratic Republic of Ethiopia         Ethiopia  ...   
1          TN                      Republic of Tunisia          Tunisia  ...   

    sectorcode source  status  supplementprojectflg  \
0  ET,BS,ES,EP   IBRD  Active                     N   
1        BZ,BS   IBRD  Active                     N   

                                              theme1  \
0      {'Name': 'Education for all', 'Percent': 100}   
1  {'Name': 'Other economic management', 'Percent...   

                                      theme_namecode themecode   totalamt  \
0      [{'name': 'Education for all', 'code': '65'}]        65  130000000   
1  [{'name': 'Other economic management', 'code':...     54,24          0   

  totalcommamt                                                url  
0    130000000  http://www.worldbank.org/projects/P129828/ethi...  
1      4700000  http://www.worldbank.org/projects/P144674?lang=en  

[2 rows x 50 columns]

****
## JSON exercise

Using data in file 'data/world_bank_projects.json' and the techniques demonstrated above,
1. Find the 10 countries with most projects
2. Find the top 10 major project themes (using column 'mjtheme_namecode')
3. In 2. above you will notice that some entries have only the code and the name is missing. Create a dataframe with the missing names filled in.

In [9]:
# Question 1.
# load project file as panda dataframe
json_df = pd.read_json('data/world_bank_projects.json')
json_df.head(3)

sector supplementprojectflg  \
0  [{'Name': 'Primary education'}, {'Name': 'Seco...                    N   
1  [{'Name': 'Public administration- Other social...                    N   
2  [{'Name': 'Rural and Inter-Urban Roads and Hig...                    Y   

  projectfinancialtype prodline  \
0                  IDA       PE   
1                OTHER       RE   
2                  IDA       PE   

                                             mjtheme  idacommamt  \
0                                [Human development]   130000000   
1  [Economic management, Social protection and ri...           0   
2  [Trade and integration, Public sector governan...     6060000   

                                  impagency  \
0                     MINISTRY OF EDUCATION   
1                       MINISTRY OF FINANCE   
2  MINISTRY OF TRANSPORT AND COMMUNICATIONS   

                                        project_name mjthemecode  \
0  Ethiopia General Education Quality Improvement...        8,11   
1          TN: DTF Social Protection Reforms Support         1,6   
2  Tuvalu Aviation Investment Project - Additiona...    5,2,11,6   

            closingdate  ...  \
0  2018-07-07T00:00:00Z  ...   
1                   NaN  ...   
2                   NaN  ...   

                                 majorsector_percent board_approval_month  \
0  [{'Percent': 46, 'Name': 'Education'}, {'Perce...             November   
1  [{'Percent': 70, 'Name': 'Public Administratio...             November   
2       [{'Percent': 100, 'Name': 'Transportation'}]             November   

                                      theme_namecode  \
0      [{'code': '65', 'name': 'Education for all'}]   
1  [{'code': '24', 'name': 'Other economic manage...   
2  [{'code': '47', 'name': 'Regional integration'...   

                               countryname  \
0  Federal Democratic Republic of Ethiopia   
1                      Republic of Tunisia   
2                                   Tuvalu   

                                                 url source  \
0  http://www.worldbank.org/projects/P129828/ethi...   IBRD   
1  http://www.worldbank.org/projects/P144674?lang=en   IBRD   
2  http://www.worldbank.org/projects/P145310?lang=en   IBRD   

  projectstatusdisplay ibrdcommamt  \
0               Active           0   
1               Active           0   
2               Active           0   

                                     sector_namecode  \
0  [{'code': 'EP', 'name': 'Primary education'}, ...   
1  [{'code': 'BS', 'name': 'Public administration...   
2  [{'code': 'TI', 'name': 'Rural and Inter-Urban...   

                                    _id  
0  {'$oid': '52b213b38594d8a2be17c780'}  
1  {'$oid': '52b213b38594d8a2be17c781'}  
2  {'$oid': '52b213b38594d8a2be17c782'}  

[3 rows x 50 columns]

In [10]:
json_df.shape

(500, 50)

In [11]:
# Check that ALL mentioned countries DO have projects whether active or closed
# 500rows of countryname and 500rows of project_name.

json_df_sub = json_df[:][['countryname','approvalfy','projectstatusdisplay','project_name']]
json_df_sub.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 4 columns):
countryname             500 non-null object
approvalfy              500 non-null int64
projectstatusdisplay    500 non-null object
project_name            500 non-null object
dtypes: int64(1), object(3)
memory usage: 15.8+ KB


In [12]:
# Count of top 10 mentioned countries
json_df['countryname'].value_counts().head(10)

Republic of Indonesia              19
People's Republic of China         19
Socialist Republic of Vietnam      17
Republic of India                  16
Republic of Yemen                  13
People's Republic of Bangladesh    12
Kingdom of Morocco                 12
Nepal                              12
Africa                             11
Republic of Mozambique             11
Name: countryname, dtype: int64

In [13]:
# Checking Africa projects
# Seems 'off' to have a Continent name, but it turns out these are projects 
# that span sub-regions.

Africa_projects = json_df[json_df['countryname']=='Africa']
Africa_projects


sector supplementprojectflg  \
45                                [{'Name': 'Health'}]                    N   
46   [{'Name': 'Telecommunications'}, {'Name': 'Inf...                    N   
51   [{'Name': 'General water, sanitation and flood...                    N   
58   [{'Name': 'General agriculture, fishing and fo...                    N   
65                            [{'Name': 'Hydropower'}]                    N   
99   [{'Name': 'Agricultural extension and research...                    Y   
167  [{'Name': 'Telecommunications'}, {'Name': 'Gen...                    N   
184  [{'Name': 'Rural and Inter-Urban Roads and Hig...                    N   
287  [{'Name': 'Agricultural extension and research...                    N   
353  [{'Name': 'General water, sanitation and flood...                    N   
449  [{'Name': 'Hydropower'}, {'Name': 'Irrigation ...                    N   

    projectfinancialtype prodline  \
45                 OTHER       RE   
46                   IDA       PE   
51                 OTHER       GE   
58                 OTHER       GM   
65                   IDA       PE   
99                   IDA       PE   
167                  IDA       PE   
184                  IDA       PE   
287                  IDA       PE   
353                OTHER       RE   
449                  IDA       PE   

                                               mjtheme  idacommamt  \
45                                 [Human development]           0   
46   [Public sector governance, Rural development, ...    22000000   
51   [Environment and natural resources management,...           0   
58   [Environment and natural resources management,...           0   
65   [Financial and private sector development, Soc...   339900000   
99   [Trade and integration, Trade and integration,...    60000000   
167  [Trade and integration, Financial and private ...    60000000   
184  [Trade and integration, Trade and integration,...   213000000   
287  [Rural development, Trade and integration, Tra...    90000000   
353  [Environment and natural resources management,...           0   
449  [Urban development, Environment and natural re...   203000000   

                                             impagency  \
45                                                WAHO   
46   RELEVANT MINISTRIES IN CHARGE OF TELECOM/ICT P...   
51                           OSS, IUCN, CILSS/AGRYHMET   
58                                       BANK EXECUTED   
65                                                 NaN   
99                                                 NaN   
167                                                NaN   
184                                           TANROADS   
287         CCARDESA AND NATIONAL GOVERNMENTS/AGENCIES   
353                              NILE BASIN INITIATIVE   
449                        NIGER BASIN AUTHORITY (NBA)   

                                          project_name     mjthemecode  \
45   West Africa  Regional Disease Surveillance Cap...            8,10   
46   RCIP4 - Regional Communications Infrastructure...        2,10,4,4   
51   Building Resilience through Innovation, Commun...  11,11,11,11,11   
58   Fighting against wildlife poaching and illegal...           11,11   
65         Regional Rusumo Falls Hydroelectric Project        4,7,10,5   
99   Additional Financing-West Africa Agricultural ...        5,5,2,10   
167  West Africa Regional Communications Infrastruc...         5,4,4,5   
184  Southern Africa Trade and Transport Facilitati...         5,5,9,8   
287  Agricultural Productivity Program for Southern...          10,5,5   
353               Nile Cooperation for Results Project         11,7,11   
449  First Part of the Second Phase of the Niger Ba...       9,11,10,5   

              closingdate  ...  \
45                    NaN  ...   
46   2018-06-30T00:00:00Z  ...   
51                    NaN  ...   
58                    NaN  ...   
65   2020-12-31T00:00:00Z  ...   
99                    N

In [14]:
# Question 2.
# Count of top 10 major project themes

# import json as string
json_string = json.load((open('data/world_bank_projects.json')))

In [66]:
# use normalization to create tables for 'mjtheme_namecode'

json_mjtheme = json_normalize(json_string,'mjtheme_namecode')
json_mjtheme.head(10) # check top 10rows

code                                          name
0    8                             Human development
1   11                                              
2    1                           Economic management
3    6         Social protection and risk management
4    5                         Trade and integration
5    2                      Public sector governance
6   11  Environment and natural resources management
7    6         Social protection and risk management
8    7                   Social dev/gender/inclusion
9    7                   Social dev/gender/inclusion

In [67]:
json_mjtheme.name.value_counts().head(10)# re-organize and display top 10 project themes

Environment and natural resources management    223
Rural development                               202
Human development                               197
Public sector governance                        184
Social protection and risk management           158
Financial and private sector development        130
                                                122
Social dev/gender/inclusion                     119
Trade and integration                            72
Urban development                                47
Name: name, dtype: int64

In [68]:
json_mjtheme.sort_values(['code', 'name']).head(10)

code                 name
212     1                     
363     1                     
1024    1                     
1114    1                     
1437    1                     
2       1  Economic management
88      1  Economic management
175     1  Economic management
204     1  Economic management
205     1  Economic management

In [69]:
# Part 3
json_mjtheme

code                                      name
0       8                         Human development
1      11                                          
2       1                       Economic management
3       6     Social protection and risk management
4       5                     Trade and integration
...   ...                                       ...
1494   10                         Rural development
1495    9                         Urban development
1496    8                         Human development
1497    5                     Trade and integration
1498    4  Financial and private sector development

[1499 rows x 2 columns]

In [70]:
#sort dataframe and place Nans at the top
json_mjtheme = json_mjtheme.sort_values('name', na_position='first')

#Create dictionary with code, name as key, value pairs
code_name = dict(zip(json_mjtheme['code'], json_mjtheme['name']))

code_name

{'8': 'Human development',
 '10': 'Rural development',
 '11': 'Environment and natural resources management',
 '4': 'Financial and private sector development',
 '7': 'Social dev/gender/inclusion',
 '2': 'Public sector governance',
 '6': 'Social protection and risk management',
 '5': 'Trade and integration',
 '3': 'Rule of law',
 '9': 'Urban development',
 '1': 'Economic management'}

In [71]:
# make copy of mjtheme
json_mjtheme_replaced = json_mjtheme.copy()

In [72]:
# Replace missing 'name' data using the dictionary key, value pairs:

json_mjtheme_replaced['name'] = json_mjtheme_replaced['code'].map(code_name)
json_mjtheme_replaced.iloc[1000:1110]

code                         name
1122   10            Rural development
1336   10            Rural development
1123   10            Rural development
195    10            Rural development
1226   10            Rural development
...   ...                          ...
224     7  Social dev/gender/inclusion
1349    7  Social dev/gender/inclusion
545     7  Social dev/gender/inclusion
890     7  Social dev/gender/inclusion
1346    7  Social dev/gender/inclusion

[110 rows x 2 columns]

In [73]:
# Check presence of Nan values
json_mjtheme_replaced['name'].isnull().value_counts()

False    1499
Name: name, dtype: int64

# Summary:

### Part 1: 10 countries with most projects, are:

People's Republic of China         19; 
Republic of Indonesia              19; 
Socialist Republic of Vietnam      17;
Republic of India                  16;
Republic of Yemen                  13;
Kingdom of Morocco                 12;
Nepal                              12;
People's Republic of Bangladesh    12;
Republic of Mozambique             11;
Africa                             11.



### Part 2:Find the top 10 major project themes (using column 'mjtheme_namecode')

Environment and natural resources management    223;
Rural development                               202;
Human development                               197;
Public sector governance                        184;
Social protection and risk management           158;
Financial and private sector development        130;
"empty cell"                                    122;
Social dev/gender/inclusion                     119;
Trade and integration                            72;
Urban development                                47.


### Part3: In 2. above, create a dataframe with the missing names filled in.
mjtheme_namecode has 2 columns. The 'code' column was complete, but the 'name' column had empty cells that were tackled by:
    1. sorting the dataframe such that the Nan values were at the top.
    2. Creating a dictionary with the 'code', 'name' as key/vale pairs.
    3. Using the dictionary to map the missing name values using the dictionary.
